### PR TITLE
Support new node types for optionals

### PR DESCRIPTION
**Pull request by xixixao**
Once you want to bump babylon, you'll need something like this. The node types have changed in https://github.com/babel/babel/pull/7288 in Feb.

### CODE REVIEW

File: src/language-js/printer-estree.js [hunk 1]
```diff
@@ -4122,12 +4135,9 @@ function printClass(path, options, print) {
 
 function printOptionalToken(path) {
   const node = path.getValue();
-  if (!node.optional) {
-    return "";
-  }
```
Comment (xixixao) [inline src/language-js/printer-estree.js]: I guess this should stay for other parsers

File: src/language-js/printer-estree.js [hunk 2]
```diff
@@ -90,12 +90,16 @@ function genericPrint(path, options, printPath, args) {
         node.type !== "ClassMethod" &&
         (decorator.type === "Identifier" ||
           decorator.type === "MemberExpression" ||
+          decorator.type === "OptionalMemberExpression" ||
           (decorator.type === "CallExpression" &&
+            decorator.type === "OptionalCallExpression" &&
```
Comment (j-f1) [inline src/language-js/printer-estree.js]: This won’t work — the node can’t be both types at the same time.

File: src/language-js/printer-estree.js (hunk 1 above)
Comment (duailibe) [reply src/language-js/printer-estree.js]: Yeah there are other types of optional nodes like optional class properties in TypeScript.

**Comment by duailibe** [comment]
We'll need this when upgrade Flow too.

@j-f1 We can't add any tests for this until we upgrade Flow and/or Babylon (which we might no be able to upgrade because of Node 4 support).

**Comment by xixixao** [comment]
I suggest you guys add the newer parsers as separate (not sure if real option or just for tests), if you want to support Node 4. I agree this should go out with exercised tests.

**Comment by vjeux** [comment]
If babel now requires node 4, we should take the plunge... It's sad but I believe it now passes the threshold of being too hard to maintain.

**Pull request closed without merging**